### DISCUSSION TITLE
How to read template literals from js?

### DISCUSSION THREAD

**Discussion by oovm** [wasm-bindgen/wasm-bindgen#2996] [Q&A] [answered]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

- rs side

```rs
#[wasm_bindgen]
pub fn parsing(template: ?) -> Result<Self, JsError> {
    // some parsing progress
}
```

- js side

```js
parsing`template {raw} text`
```

**Comment by Liamolucko** [comment, marked as answer]
Looking at that MDN page, the signature of a tag functions seems to be:

```ts
function tag(strings: string[] & { raw: string[] }, ...args: unknown[]): unknown;
```

The variadic `args` argument is the tricky part, since exporting functions with variadic arguments isn't supported as of wasm-bindgen 0.2.81. However, support was added in #2954, and will be there in the next release.

So, using a git version of wasm-bindgen, a tag function can be written like this:

```rust
use js_sys::Array;
use wasm_bindgen::prelude::*;

#[wasm_bindgen(variadic)]
pub fn tag(strings: Array, args: Array) -> Array {
    // ...
}
```

`#[wasm_bindgen(variadic)]` makes the last argument of the function an array of all the arguments passed after all the non-variadic arguments (in this case, `args`).

And, if you need to access the `raw` property of the first argument, you can define a subtype of `Array` which includes it:

```rust
use js_sys::Array;
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(extends = Array)]
    pub type Strings;

    #[wasm_bindgen(method, getter)]
    pub fn raw(this: &Strings) -> Array;
}

#[wasm_bindgen(variadic)]
pub fn tag(strings: Strings, args: Array) -> Array {
    // ...
}
```

Again though, **this support hasn't been released yet**, so to use it you'll need a git version of wasm-bindgen.

To do that, you have to add this to your `Cargo.toml`:

```toml
[patch.crates-io]
wasm-bindgen = { git = "https://github.com/rustwasm/wasm-bindgen" }
```
And install the git version of the CLI (note that this will be incompatible with anything compiled with non-git versions of wasm-bindgen):

```sh
cargo install --git=https://github.com/rustwasm/wasm-bindgen
```
